cmd/fs-v1.go

```diff
@@ -452,13 +452,17 @@ func (fs fsObjects) GetObject(bucket, object string, offset int64, length int64,
 // getObjectInfo - wrapper for reading object metadata and constructs ObjectInfo.
 func (fs fsObjects) getObjectInfo(bucket, object string) (oi ObjectInfo, e error) {
 	fsMeta := fsMetaV1{}
-	if hasSuffix(object, slashSeparator) {
-		// Directory call needs to arrive with object ending with "/".
-		fi, err := fsStatDir(pathJoin(fs.fsPath, bucket, object))
-		if err != nil {
-			return oi, toObjectErr(err, bucket, object)
+	fi, err := fsStatDir(pathJoin(fs.fsPath, bucket, object))
+	if err != nil && errorCause(err) != errFileAccessDenied {
+		return oi, toObjectErr(err, bucket, object)
+	}
+	if fi != nil {
+		// If file found and request was with object ending with "/", consider it
+		// as directory and return object info
+		if hasSuffix(object, slashSeparator) {
+			return fsMeta.ToObjectInfo(bucket, object, fi), nil
 		}
-		return fsMeta.ToObjectInfo(bucket, object, fi), nil
+		return oi, toObjectErr(errFileNotFound, bucket, object)
```

Comment:
Returning an `errFileNotFound` here is unhelpful, actually: it's now impossible to distinguish between an object that does not exist and an object which is actually a directory without sending also a request with a trailing `/`.

Would you consider changing that? Should I create an issue about it?

Reply:
I see your comment @Itizir :) 

I think we can pass on the error like we used to do, but let me check with the team as well.

Reply:
Ah good, good.
Cool, sure thing. And thanks for the quick reply!

Reply:
Directory is not a object as such. So this code is correct and it is a correct requirement to have a special interpretation for / . This is how AWS S3 allows you to create directories.

The current code is correct. If you are writing a new application then you should avoid creating directories in the first place.  Just upload the object with right prefix parent directories will be created automatically.

The current code is meant for legacy application requirements.

Reply:
Sorry, but I'm a bit confused by your answer... I'm also no S3 expert, so won't be fully familiar with its conventions: you may have to teach me some basics if I got it all wrong. And it you're basically saying minio is fully consistent with the S3 API I suppose I would have to go complain to Amazon... ;)
Anyway, bear with me for a minute!

I more or less understand that object names are not formally representing paths in a directory structure, but I thought this meant `/` is in fact _not_ special.
Pursuing this logic all the way, wouldn't that mean you could in principle have both an object called `someprefix/objectname` and another called simply `someprefix`? But you can't do this with minio...

I also understand that I shouldn't manually go an mess with the underlying directory structure, but if you're saying the burden should be on the user to make sure to use only consistent and compatible prefixes and object names, then there should be a way to check which prefixes/objects can be used at a given time.

Regardless, I'm not sure how it justifies the current behaviour of `StatObject`. Let me summarise (assuming there is one object in the bucket called `someprefix/objectname`).

What seems consistent to me:
- a query for `someprefix/objectname`: returns no error, since the object exists
- `someprefix/objectname/` (or anything starting with that): access denied, and indeed wouldn't be able to put an object `someprefix/objectname/otherthing`; although I guess I misunderstood the point of this MR, which sounded like this should return the same as above
- `someprefix/object` and `someprefix/object/`: no such key, and indeed could use this as object or prefix, respectively

What puzzles me a bit:
- `someprefix`: no such key; there is indeed no such object, however creating one with that name is not allowed
- `someprefix/`: no error, even though this is not an existing object

That's about it. Please do let me know if I'm just being obtuse, or if indeed I may have a point...